File: task1.py
def calculate_product_total(product_name, price, quantity, is_gift_wrapped):
    product_total = price * quantity
    if is_gift_wrapped:
        product_total += quantity * 1 
    print(f"{product_name}: Quantity - {quantity}, Total - ${product_total}")
    return product_total

def discounts(cart_total, product_quantities):
    discounts = {
        "flat_10_discount": 10 if cart_total > 200 else 0,
        "bulk_5_discount": 0,
        "bulk_10_discount": 0,
        "tiered_50_discount": 0
    }

    for product_quantity in product_quantities.values():
        if product_quantity > 10:
            discounts["bulk_5_discount"] = 5
        if sum(product_quantities.values()) > 20:
            discounts["bulk_10_discount"] = 10
        if sum(product_quantities.values()) > 30 and any(quantity > 15 for quantity in product_quantities.values()):
            discounts["tiered_50_discount"] = 50
            break

    applicable_discount = max(discounts, key=discounts.get)
    discount_amount = (cart_total * discounts[applicable_discount]) / 100
    return applicable_discount, discount_amount

def shipping_fee(total_quantity):
    return (total_quantity // 10) * 5

def main():
    products = {
        "Product A": 20,
        "Product B": 40,
        "Product C": 50
    }
    product_quantities = {}
    total = 0

    for product_name, price in products.items():
        quantity = int(input(f"Enter the quantity of {product_name}: "))
        is_gift_wrapped = input(f"Do you want to wrap {product_name} as a gift? (yes/no): ").lower() == "yes"
        total += calculate_product_total(product_name, price, quantity, is_gift_wrapped)
        product_quantities[product_name] = quantity

    cart_total = total
    discount_name, discount_amount = discounts(cart_total, product_quantities)
    shipping_fee = shipping_fee(sum(product_quantities.values()))
    total -= discount_amount
    total += shipping_fee

    print(f"\nSubtotal: ${cart_total}")
    print(f"Discount Applied: {discount_name} - ${discount_amount}")
    print(f"Shipping Fee: ${shipping_fee}")
    print(f"Total: ${total}")

if __name__ == "__main__":
    main()

File: task1.js

function ProductTotal(productName, price, quantity, isGiftWrapped) {
    let productTotal = price * quantity;
    if (isGiftWrapped) {
        productTotal += quantity * 1; // Gift wrap fee is $1 per unit
    }
    console.log(`${productName}: Quantity - ${quantity}, Total - $${productTotal}`);
    return productTotal;
}

// Function to apply discounts based on the rules provided
function Discounts(cartTotal, productQuantities) {
    let discounts = {
        "flat_10_discount": cartTotal > 200 ? 10 : 0,
        "bulk_5_discount": 0,
        "bulk_10_discount": 0,
        "tiered_50_discount": 0
    };

    for (let quantity of Object.values(productQuantities)) {
        if (quantity > 10) {
            discounts["bulk_5_discount"] = 5;
        }
    }

    if (Object.values(productQuantities).reduce((a, b) => a + b) > 20) {
        discounts["bulk_10_discount"] = 10;
    }

    if (Object.values(productQuantities).reduce((a, b) => a + b) > 30 && Object.values(productQuantities).some(quantity => quantity > 15)) {
        discounts["tiered_50_discount"] = 50;
    }

    let applicableDiscount = Object.keys(discounts).reduce((a, b) => discounts[a] > discounts[b] ? a : b);
    let discountAmount = (cartTotal * discounts[applicableDiscount]) / 100;
    return [applicableDiscount, discountAmount];
}

// Function to calculate shipping fee
function calculateShippingFee(totalQuantity) {
    return Math.floor(totalQuantity / 10) * 5;
}

// Function to get user input and calculate the total
function main() {
    const products = {
        "Product A": 20,
        "Product B": 40,
        "Product C": 50
    };
    let productQuantities = {};
    let total = 0;

    for (let [productName, price] of Object.entries(products)) {
        let quantity = parseInt(prompt(`Enter the quantity of ${productName}: `));
        let isGiftWrapped = prompt(`Do you want to wrap ${productName} as a gift? (yes/no): `).toLowerCase() === "yes";
        total += ProductTotal(productName, price, quantity, isGiftWrapped);
        productQuantities[productName] = quantity;
    }

    let cartTotal = total;
    let [discountName, discountAmount] = Discounts(cartTotal, productQuantities);
    let shippingFee = calculateShippingFee(Object.values(productQuantities).reduce((a, b) => a + b));
    total -= discountAmount;
    total += shippingFee;

    console.log(`\nSubtotal: $${cartTotal}`);
    console.log(`Discount Applied: ${discountName} - $${discountAmount}`);
    console.log(`Shipping Fee: $${shippingFee}`);
    console.log(`Total: $${total}`);
}

main();
